File: chat-uikit/src/main/java/io/agora/chat/uikit/activities/EaseShowNormalFileActivity.java
package io.agora.chat.uikit.activities;

import android.os.Bundle;
import android.widget.ProgressBar;
import android.widget.Toast;

import io.agora.CallBack;
import io.agora.Error;
import io.agora.chat.ChatClient;
import io.agora.chat.ChatMessage;
import io.agora.chat.FileMessageBody;
import io.agora.chat.uikit.R;
import io.agora.chat.uikit.base.EaseBaseActivity;
import io.agora.chat.uikit.utils.EaseCompat;


public class EaseShowNormalFileActivity extends EaseBaseActivity {
    private static final String TAG = EaseShowNormalFileActivity.class.getSimpleName();
	private ProgressBar progressBar;

    @Override
    public void setActivityTheme() {
        super.setActivityTheme();
    }

    @Override
	protected void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		setContentView(R.layout.ease_activity_show_file);
        setFitSystemForTheme(true, R.color.transparent, true);
		progressBar = (ProgressBar) findViewById(R.id.progressBar);

		final ChatMessage message = getIntent().getParcelableExtra("msg");
        if (!(message.getBody() instanceof FileMessageBody)) {
            Toast.makeText(EaseShowNormalFileActivity.this, "Unsupported message body", Toast.LENGTH_SHORT).show();
            finish();
            return;
        }
        message.setMessageStatusCallback(new CallBack() {
            @Override
            public void onSuccess() {
                runOnUiThread(new Runnable() {
                    public void run() {
                        EaseCompat.openFile(EaseShowNormalFileActivity.this,
                                ((FileMessageBody) message.getBody()).getLocalUri());
                        finish();
                    }
                });

            }

            @Override
            public void onError(final int code, final String error) {
                runOnUiThread(new Runnable() {
                    public void run() {
                        EaseCompat.deleteFile(EaseShowNormalFileActivity.this, ((FileMessageBody) message.getBody()).getLocalUri());
                        String str4 = getResources().getString(R.string.ease_failed_to_download_file);
                        if (code == Error.FILE_NOT_FOUND) {
                            str4 = getResources().getString(R.string.ease_file_expired);
                        }
                        Toast.makeText(getApplicationContext(), str4+message, Toast.LENGTH_SHORT).show();
                        finish();
                    }
                });
            }

            @Override
            public void onProgress(final int progress, String status) {
                runOnUiThread(new Runnable() {
                    public void run() {
                        progressBar.setProgress(progress);
                    }
                });
            }
        });
        ChatClient.getInstance().chatManager().downloadAttachment(message);
	}
}

File: ease-im-kit/src/main/kotlin/io/agora/uikit/feature/chat/activities/EaseShowNormalFileActivity.kt
package io.agora.uikit.feature.chat.activities

import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.view.LayoutInflater
import android.widget.Toast
import io.agora.uikit.EaseIM
import io.agora.uikit.R
import io.agora.uikit.base.EaseBaseActivity
import io.agora.uikit.common.ChatClient
import io.agora.uikit.common.ChatError
import io.agora.uikit.common.ChatFileMessageBody
import io.agora.uikit.common.ChatMessage
import io.agora.uikit.common.extensions.hasRoute
import io.agora.uikit.common.extensions.mainScope
import io.agora.uikit.common.impl.CallbackImpl
import io.agora.uikit.common.utils.EaseCompat
import io.agora.uikit.databinding.EaseActivityShowFileBinding
import kotlinx.coroutines.launch

class EaseShowNormalFileActivity : EaseBaseActivity<EaseActivityShowFileBinding>() {
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        val message: ChatMessage? = intent.getParcelableExtra("msg")
        if (message?.body !is ChatFileMessageBody) {
            Toast.makeText(
                this@EaseShowNormalFileActivity,
                "Unsupported message body",
                Toast.LENGTH_SHORT
            ).show()
            finish()
            return
        }
        message.setMessageStatusCallback(CallbackImpl(
            onSuccess = {
                mainScope().launch{
                    EaseCompat.openFile(
                        mContext,
                        (message.body as ChatFileMessageBody).localUri
                    )
                    finish()
                }
            },
            onError = { code, error ->
                mainScope().launch {
                    EaseCompat.deleteFile(
                        mContext,
                        (message.body as ChatFileMessageBody).localUri
                    )
                    var str4: String =
                        resources.getString(R.string.ease_failed_to_download_file)
                    if (code == ChatError.FILE_NOT_FOUND) {
                        str4 = resources.getString(R.string.ease_file_expired)
                    }
                    Toast.makeText(mContext, str4 + message, Toast.LENGTH_SHORT)
                        .show()
                    finish()
                }
            },
            onProgress = { progress ->
                mainScope().launch {
                    this@EaseShowNormalFileActivity.binding.progressBar.progress = progress
                }
            }
        ))
        ChatClient.getInstance().chatManager().downloadAttachment(message)
    }

    override fun getViewBinding(inflater: LayoutInflater): EaseActivityShowFileBinding {
        return EaseActivityShowFileBinding.inflate(inflater)
    }


    companion object {
        private val TAG = EaseShowNormalFileActivity::class.java.simpleName
        fun actionStart(context: Context, message: ChatMessage?) {
            val intent = Intent(context, EaseShowNormalFileActivity::class.java)
            intent.putExtra("msg", message)
            EaseIM.getCustomActivityRoute()?.getActivityRoute(intent.clone() as Intent)?.let {
                if (it.hasRoute()) {
                    context.startActivity(it)
                    return
                }
            }
            context.startActivity(intent)
        }
    }
}